### DISCUSSION TITLE
Accessing previous state within StateProvider.

### DISCUSSION THREAD

**Discussion by TreyThomas93** [rrousselGit/riverpod#1790] [General]
I was wondering if there was a way to access the previous state when calling a StateProvider. I attempt to use **ref.controller.state**, but I end up with a ***The type of 'StateProvider' cannot be inferred because it depends on itself through the cycle: [StateProvider]*** error.

**Comment by rrousselGit** [comment]
The error sounds like you didn't specify the generic type for your provider.

Do:

```dart
final provider = STateProvider<T>(...);
```

not:


```dart
final provider = STateProvider(...);
```

**Reply by TreyThomas93** [reply]
@rrousselGit I did add a type to the StateProvider: 
```dart
StateProvider<List<Note>>
```

**Comment by g-30** [comment]
Support similar request for `StateNotifierProvider` here: https://github.com/rrousselGit/riverpod/discussions/1259#discussioncomment-4780284
